node/src/main/scala/com/wavesplatform/state/diffs/invoke/InvokeScriptDiff.scala

```diff
@@ -188,18 +190,47 @@ object InvokeScriptDiff {
                 )
               )
 
-            (actionsDiff, evaluated) <- scriptResult match {
+            (actionsDiff, evaluated, remainingActions1, remainingData1) <- scriptResult match {
               case ScriptResultV3(dataItems, transfers, unusedComplexity) =>
-                doProcessActions(dataItems ::: transfers, unusedComplexity).map((_, unit))
-              case ScriptResultV4(actions, unusedComplexity, ret) => doProcessActions(actions, unusedComplexity).map((_, ret))
-              case _: IncompleteResult if limitedExecution        => doProcessActions(Nil, 0).map((_, unit))
+                val dataCount = dataItems.length
+                if(dataCount > avaliableData) {
+                  val usedComplexity = remainingComplexity - unusedComplexity
+                  val error          = FailedTransactionError.dAppExecution("Stored data count limit is exceeded", usedComplexity, log)
+                  traced(error.asLeft[(Diff, EVALUATED, Int, Int)])
+                } else {
+                  val actionsCount = transfers.length
+                  if(actionsCount > avaliableActions) {
+                    val usedComplexity = remainingComplexity - unusedComplexity
+                    val error          = FailedTransactionError.dAppExecution("Actions count limit is exceeded", usedComplexity, log)
+                    traced(error.asLeft[(Diff, EVALUATED, Int, Int)])
+                  } else {
+                    doProcessActions(dataItems ::: transfers, unusedComplexity).map((_, unit, avaliableActions - actionsCount, avaliableData - dataCount))
+                  }
+                }
+              case ScriptResultV4(actions, unusedComplexity, ret) =>
```

Comment:
Code below seems as duplication from `case ScriptResultV3` block. Please, extract to method and reuse if it possible